Add EngineSource, EngineManager, and Mayavi2 plugin panel

kitchoi opened the pull request:
Main features:
- [x] `EngineSource` for adapting CUDS from a simphony engine (analogous to `CUDSFileSource`) 
- [x] `EngineManager` for selecting one modeling engines out of many (TraitsUI)
- [x] `EngineManagerStandalone` for visualising datasets from engine, running and animating the results (GUI-free)
- [x] `EngineManagerStandaloneUI` a GUI version of `EngineManagerStandalone`
- [x] `EngineManagerMayavi2` as a panel to sit inside Mayavi2
- [x] `EngineManagerMayavi2Plugin` a envisage plugin for managing `EngineManagerMayavi2`
- [x] Added tests (except for `EngineManagerMayavi2`)
- [x] Added documentation
- [ ] Tests for `EngineManagerMayavi2`

Supporting elements:
- [x] `AddEngineSourceToMayavi`, `AddSourcePanel`, `RunAndAnimate` and `RunAndAnimatePanel` (for reusability)
- [x] `TabbedPanelCollection` collects multiple panels and displays as tabs
- [x] `default_modules` for setting default mayavi modules given the point_scalars_name/... (also see changes in `show.py`)

dpinte commented on simphony_mayavi/modules/default_module.py:
```diff
@@ -0,0 +1,85 @@
+        return modules
+    else:
+        message = "Unknown scalar/vector data for setting default module"
+        logging.warning(message)
```
@kitchoi the typical way of using logging is to get a logger from the logging. If you don't target a specific logger, you can get one for this module:

```
logger = logging.getLogger(__name__)
```

then only use that logger instance for the calls:

```
logger.warning(xxx)
```

dpinte commented on simphony_mayavi/plugins/add_source_panel.py:
```diff
@@ -0,0 +1,177 @@
+    # Public methods
+    # --------------------------------------------------
+
+    def __init__(self, engine_name, engine, mayavi_engine):
```
you typically avoid doing this with `HasTraits` class. In this case, you break the traits initialization and event system. 

You can remove the `__init__` method and use the standard trait intialization by keyword:

```
AddSourcePanel(engine=engine, engine_name=engine_name, mayavi_engine=mayavi_engine)
```

If you really, really need an `__init__`  method, you should calls `HasTraits.__init__(self, **kwargs)` to properly intialize the `HasTraits` class (with the appropriate definitions of the `kwargs`)

dpinte commented on simphony_mayavi/plugins/add_source_panel.py:
```diff
@@ -0,0 +1,177 @@
+    label = "Add to Mayavi"
+
+    # Buttons for the UI
+    _add_dataset = Button("+")
```
I tend to prefer using a `Event` trait with a `ButtonEditor` in the UI

dpinte commented on simphony_mayavi/plugins/engine_manager.py:
```diff
@@ -0,0 +1,95 @@
+    _engine_names = ListStr
+
+    # Selected engine (overloaded BasicPanel)
+    engine = Property
```
for the property, you tycally pass the type of the property and what it depends on

dpinte commented on simphony_mayavi/plugins/engine_manager_mayavi2.py:
```diff
@@ -0,0 +1,128 @@
+    def _service_offers_default(self):
+        """ Trait initializer. """
+        worker_service_offer = ServiceOffer(
+            protocol=('simphony_mayavi.plugins.engine_manager_mayavi2.'
```
it's better to keep the protocol name as one single string (even if too long) and add a `# noqa`. If not, searching for that string will always fail on those entries.

dpinte commented on simphony_mayavi/plugins/engine_manager_standalone_ui.py:
```diff
@@ -0,0 +1,65 @@
+            Group(Item("panels", style="custom", show_label=False))),
+        title="Engine Manager")
+
+    def __init__(self, engine_name, engine, mayavi_engine=None):
```
if `EngineManager` is a `HasTraits` class, this `__init__` is broken

dpinte commented on simphony_mayavi/plugins/run_and_animate_panel.py:
```diff
@@ -0,0 +1,137 @@
+    # Public methods
+    # ----------------------------------------------------------
+
+    def __init__(self, engine, mayavi_engine):
```
`HasTraits.__init__`

dpinte commented on simphony_mayavi/plugins/tests/test_add_engine_source_to_mayavi.py:
```diff
@@ -0,0 +1,101 @@
+from simphony_mayavi.sources.tests import testing_utils
+
+
+class TestAddEngineSourceToMayavi(unittest.TestCase):
```
We should talk about testing traits class during our call today

dpinte commented:
@kitchoi great work! See my comments in the code

kitchoi commented:
@dpinte Thanks for reviewing the code!
I have:
- added trait's tests (`traits.testing.UnittestTools`)  for `EngineManager`, `AddSourcePanel` and `EngineManagerStandaloneUI` (`AddEngineSourceToMayavi` and `RunAndAnimate` are not subclasses of `HasTrait`)
- kept the `__init__` for `EngineManagerStandaloneUI` for it to be consistent with `EngineManagerStandalone`
- `__init__` in `AddSourcePanel`, `RunAndAnimatePanel` and `TabbedPanelCollection` are removed.
- replaced `logging.warn/info/..` with `logger.warn/info/...`
- added more tests for `default_module`

Forgot to ask, is it possible to test `EngineManagerMayavi2` and `EngineManagerMayavi2Plugin` as it is being loaded by envisage?

kitchoi commented:
@dpinte is this good to merge?

As for loading engine wrappers from *.py file or from default factory function, I am planning to put this into a separate PR:
![add_engine_panel](https://cloud.githubusercontent.com/assets/3673984/12330809/ce8f3f82-bade-11e5-8681-2fe5e894ba9c.png)

dpinte commented:
@kitchoi LGTM. And a separate PR for the latest additions is definitely a good idea

kitchoi commented:
Thanks!
